### DISCUSSION TITLE
Cannot build k2. Can anyone who successfully built the environment share a mirror?(a docker file etc)

### DISCUSSION THREAD

**Discussion by CopyNinja1999** [lifeiteng/valle#35] [General]
followed the instruction and got this error:
`-- CMAKE_VERSION: 3.10.2
-- No NVCC detected. Disable CUDA support
-- Enabled languages: CXX
-- No CMAKE_BUILD_TYPE given, default to Release
-- Set K2_ENABLE_NVTX to OFF since K2_WITH_CUDA is OFF
-- K2_OS: Ubuntu 18.04.4 LTS
-- C++ Standard version: 14
-- Could NOT find Valgrind (missing: Valgrind_INCLUDE_DIR Valgrind_EXECUTABLE) 
-- Downloading pybind11 from https://github.com/pybind/pybind11/archive/5bc0943ed96836f46489f53961f6c438d2935357.zip
-- pybind11 is downloaded to /home/zhangzheyu/valle/k2/build/temp.linux-x86_64-cpython-310/_deps/pybind11-src
-- pybind11 v2.11.0 dev1
-- Found PythonInterp: /home/zhangzheyu/miniconda3/envs/lft/bin/python3 (found suitable version "3.10.9", minimum required is "3.6") 
-- Found PythonLibs: /home/zhangzheyu/miniconda3/envs/lft/lib/libpython3.10.so
-- Python executable: /home/zhangzheyu/miniconda3/envs/lft/bin/python3
-- The CUDA compiler identification is unknown
CMake Error at /home/zhangzheyu/miniconda3/envs/lft/lib/python3.10/site-packages/torch/share/cmake/Caffe2/public/cuda.cmake:47 (enable_language):
  No CMAKE_CUDA_COMPILER could be found.

  Tell CMake where to find the compiler by setting either the environment
  variable "CUDACXX" or the CMake cache entry CMAKE_CUDA_COMPILER to the full
  path to the compiler, or to the compiler name if it is in the PATH.
Call Stack (most recent call first):
  /home/zhangzheyu/miniconda3/envs/lft/lib/python3.10/site-packages/torch/share/cmake/Caffe2/Caffe2Config.cmake:92 (include)
  /home/zhangzheyu/miniconda3/envs/lft/lib/python3.10/site-packages/torch/share/cmake/Torch/TorchConfig.cmake:68 (find_package)
  cmake/torch.cmake:11 (find_package)
  CMakeLists.txt:292 (include)

-- Configuring incomplete, errors occurred!
See also "/home/zhangzheyu/valle/k2/build/temp.linux-x86_64-cpython-310/CMakeFiles/CMakeOutput.log".
See also "/home/zhangzheyu/valle/k2/build/temp.linux-x86_64-cpython-310/CMakeFiles/CMakeError.log".
cat: k2/csrc/version.h: No such file or directory
make: *** No rule to make target 'install'.  Stop.
Traceback (most recent call last):
 ……skip……
  File "/home/zhangzheyu/valle/k2/setup.py", line 206, in build_extension
    raise Exception("Failed to build k2")
Exception: Failed to build k2`

**Comment by lifeiteng** [comment]
ERROR `No CMAKE_CUDA_COMPILER could be found.`

@CopyNinja1999 you can turn off CUDA, We didn't use k2 (but it is used in icefall)

```
cd k2
export K2_MAKE_ARGS="-j12"
export K2_CMAKE_ARGS="-DK2_WITH_CUDA=OFF"
python setup.py install
cd -
```

**Reply by CopyNinja1999** [reply]
The "export" instructions are followed- but still not working.

**Comment by MisakaMikoto96** [comment]
> followed the instruction and got this error: `-- CMAKE_VERSION: 3.10.2 -- No NVCC detected. Disable CUDA support -- Enabled languages: CXX -- No CMAKE_BUILD_TYPE given, default to Release -- Set K2_ENABLE_NVTX to OFF since K2_WITH_CUDA is OFF -- K2_OS: Ubuntu 18.04.4 LTS -- C++ Standard version: 14 -- Could NOT find Valgrind (missing: Valgrind_INCLUDE_DIR Valgrind_EXECUTABLE) -- Downloading pybind11 from https://github.com/pybind/pybind11/archive/5bc0943ed96836f46489f53961f6c438d2935357.zip -- pybind11 is downloaded to /home/zhangzheyu/valle/k2/build/temp.linux-x86_64-cpython-310/_deps/pybind11-src -- pybind11 v2.11.0 dev1 -- Found PythonInterp: /home/zhangzheyu/miniconda3/envs/lft/bin/python3 (found suitable version "3.10.9", minimum required is "3.6") -- Found PythonLibs: /home/zhangzheyu/miniconda3/envs/lft/lib/libpython3.10.so -- Python executable: /home/zhangzheyu/miniconda3/envs/lft/bin/python3 -- The CUDA compiler identification is unknown CMake Error at /home/zhangzheyu/miniconda3/envs/lft/lib/python3.10/site-packages/torch/share/cmake/Caffe2/public/cuda.cmake:47 (enable_language): No CMAKE_CUDA_COMPILER could be found.
> 
> Tell CMake where to find the compiler by setting either the environment variable "CUDACXX" or the CMake cache entry CMAKE_CUDA_COMPILER to the full path to the compiler, or to the compiler name if it is in the PATH. Call Stack (most recent call first): /home/zhangzheyu/miniconda3/envs/lft/lib/python3.10/site-packages/torch/share/cmake/Caffe2/Caffe2Config.cmake:92 (include) /home/zhangzheyu/miniconda3/envs/lft/lib/python3.10/site-packages/torch/share/cmake/Torch/TorchConfig.cmake:68 (find_package) cmake/torch.cmake:11 (find_package) CMakeLists.txt:292 (include)
> 
> -- Configuring incomplete, errors occurred! See also "/home/zhangzheyu/valle/k2/build/temp.linux-x86_64-cpython-310/CMakeFiles/CMakeOutput.log". See also "/home/zhangzheyu/valle/k2/build/temp.linux-x86_64-cpython-310/CMakeFiles/CMakeError.log". cat: k2/csrc/version.h: No such file or directory make: *** No rule to make target 'install'. Stop. Traceback (most recent call last): ……skip…… File "/home/zhangzheyu/valle/k2/setup.py", line 206, in build_extension raise Exception("Failed to build k2") Exception: Failed to build k2`


there has some problem on installing k2.
but you could follow this issue and I succeed.

https://github.com/k2-fsa/k2/issues/1158

**Comment by lifeiteng** [comment]
export PATH=/usr/local/cuda/bin:${PATH}

**Comment by lifeiteng** [comment]
you should config this in your ~/.bashrc or ~/.zshrc

```
export PATH=/usr/local/cuda/bin:${PATH}
export LD_LIBRARY_PATH=/usr/local/cuda/lib64:$LD_LIBRARY_PATH
export LD_LIBRARY_PATH=/usr/local/cuda/include:$LD_LIBRARY_PATH
```

**Comment by TsungChiCheng** [comment]
Hi,
I skipped the source building of k2 by downloading the pre-compiled whl file from [here](https://k2-fsa.github.io/k2/installation/pre-compiled-cuda-wheels-linux/index.html) and find the required one which meet my environment.
Hope this can help.

**Reply by DoloresHaaze** [reply]
This worked for me, thanks :D
